kong/kong.lua

```diff
@@ -252,6 +252,18 @@ function Kong.balancer()
   end
 end
 
+function Kong.rewrite()
+  core.rewrite.before()
+
+  for _, plugin in ipairs(singletons.loaded_plugins) do
+    if plugin.schema.no_consumer and plugin.schema.no_api then
+      plugin.handler:rewrite()
+    end
+  end
+
+  core.rewrite.after()
+end
+
```

Comment:
@thefosk That's the option 3 above, which is nice. But that is a limitation that kicks in when configuring. But you should not be using that field to select the plugins to run, you should run all plugins that are configured as global, so where `consumer_id` and `api_id` are `nil`.

Running the global plugins is a Kong responsibility
Marking a plugin with `no_api` is a plugin developer responsibility.
Those are different things.